subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/ivy/RealisedIvyModuleResolveMetadataSerializationHelper.java

```diff
@@ -57,6 +58,10 @@
 import java.util.Set;
 
 public class RealisedIvyModuleResolveMetadataSerializationHelper extends AbstractRealisedModuleResolveMetadataSerializationHelper {
+
+    private static final ImmutableSet<String> API_ELEMENTS_HIERARCHY = ImmutableSet.of("compile", "apiElements");
```

Comment:
The whole set of changes in this class should go away unless the hierarchy aspect has a meaning I missed.

Reply:
This is a good one. So the rational behind this is:

`IvyConfigurationHelper` holds all the excludes and `RealisedIvyModuleResolveMetadataSerializationHelper` makes use of `configurationHelper.filterExcludes(hierarchy)`

`readIvyConfigurations` https://github.com/gradle/gradle/blob/master/subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/ivy/RealisedIvyModuleResolveMetadataSerializationHelper.java#L99 iterates on the known configurations and for cases like `apiElements` and `runtimeElements`, there is no definition.

My idea here in order to avoid refactoring the way the excludes are retrieving, was to represent the hierarchy and then the `IvyConfigurationHelper` would know which excludes to pass: https://github.com/gradle/gradle/blob/master/subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/ivy/IvyConfigurationHelper.java#L79-L80

Do you think the code should be change in a way that the helper retrieves the excludes for a given configuration? unfortunately that might lose the hierarchy if `compile` extends something else or has something else?

what are your thoughts on this?

Reply:
With the above, I think this is a better approach then: https://github.com/gradle/gradle/pull/12480/commits/7fe2a67c7bc021b62b5e0a98d5cc001e5614b2ad

Keep track of the hierarchies and instance of constructing the hierarchy, if apiElements or runtimeElements, go and take it from the previous seen ones.

Reply:
Ok, I got it. I had missed this in my initial review.

However your solution does not work. It hardcodes the link between a base Ivy configuration and a variant added by a rule. There is nothing that prevents, though I would not recommend it, to keep classical variant names but instead derive them from exotic Ivy configurations. And in that case, the copy of excludes must work. Otherwise you could end up with a different behavior based on having or not cache component metadata rules (one way to have the realised code path exercised for real).

I have pushed a commit that fixes the issue, takes care of the cache layout version and also updates the test closer to what I had suggested.

Reply:
> to keep classical variant names but instead derive them from exotic Ivy configurations. 

Yeah, I did some of bad assumption there or limited scope.

Thank you for fixing that portion!